### PR TITLE
DOCKER-1088 sdc-docker should not modify VMs' internal docker:nfsvolumes metadata

### PR DESCRIPTION
**Pull request by misterdjules**

### CODE REVIEW

File: test/vms.volumes.test.js [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+        server_uuid: SERVER.uuid,
+        networks: [ { uuid: ADMIN_FABRIC_NETWORK.uuid } ],
+        brand: 'joyent-minimal',
+        billing_id: '00000000-0000-0000-0000-000000000000',
+        ram: 64,
+        quota: 10,
+        /*
+         * Not setting a cpu_cap here would break the ability for Triton to
+         * provision any VM with a non-null cpu_cap, since a mix of capped and
+         * cap-less VMs is not allowed by the allocation system.
```
Comment (joshwilsdon) [inline test/vms.volumes.test.js]: By setting a cap, you're also going to have problems if the system is setup for manta which has no caps. There really isn't any winning. :)

File: test/vms.volumes.test.js [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+        function (err, req, res, networks) {
+        common.ifError(t, err);
+        t.equal(res.statusCode, 200, '200 OK');
+        t.ok(networks, 'networks is set');
+        t.ok(Array.isArray(networks), 'networks is Array');
+        t.ok(networks.length === 1, '1 network found');
+
+        ADMIN_FABRIC_NETWORK = networks[0];
+        t.ok(ADMIN_FABRIC_NETWORK,
+                'Admin fabric network should have been found');
```
Comment (joshwilsdon) [inline test/vms.volumes.test.js]: again: github? Or extra spaces?

File: test/vms.volumes.test.js [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+                        function onVmCreated(createVmErr, req, res, body) {
+                            var expectedResStatusCode = 202;
+
+                            if (body) {
+                                vmProvisioningJobUuid = body.job_uuid;
+                                vmUuid = body.vm_uuid;
+                            }
+
+                            t.ifError(createVmErr,
+                                    'VM creation should not error');
```
Comment (joshwilsdon) [inline test/vms.volumes.test.js]: I'm hoping this is just github, because then we can call this "github spacing".

File: test/vms.volumes.test.js [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+                        next();
+                        return;
+                    }
+
+                    waitForValue('/jobs/' + vmProvisioningJobUuid, 'execution',
+                        'succeeded',
+                        { client: client, timeout: 4 * 60 * 1000 },
+                        function onVmProvisioned(provisionErr) {
+                            t.ifError(provisionErr,
+                                    'VM should be provisioned successfully');
```
Comment (joshwilsdon) [inline test/vms.volumes.test.js]: github spaced?

File: test/vms.volumes.test.js [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+                            next();
+                        });
+                },
+
+                function checkVolumeProvisioned(ctx, next) {
+                    waitForValue('/volumes/' + volumeUuid, 'state', 'ready',
+                        { client: client.volapi },
+                        function onVolCreated(volCreatErr) {
+                            t.ifError(volCreatErr,
+                                    'VM should be provisioned successfully');
```
Comment (joshwilsdon) [inline test/vms.volumes.test.js]: huh. I'm questioning myself.

File: test/vms.volumes.test.js [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+                        next();
+                    });
+                },
+
+                function waitForVmDeleted(ctx, next) {
+                    waitForValue('/vms/' + vmUuid, 'state', 'destroyed', {
+                        client: client
+                    }, function onVmDeleted(vmDelErr) {
+                        t.ifError(vmDelErr,
+                                'VM should have been deleted successfully');
```
Comment (joshwilsdon) [inline test/vms.volumes.test.js]: feels like too much coincidence.

File: lib/workflows/nfs-volumes/index.js [hunk 7]
```diff
@@ -83,48 +82,20 @@ function provisionNfsVolumes(job, cb) {
...
-            requiredNfsVolumesString =
-                vm.internal_metadata['docker:nfsvolumes'];
+            requiredVolumes = vm.volumes;
 
-            if (requiredNfsVolumesString === undefined) {
-                // Not need to continue, as this VM does not depend on any NFS
-                // volume.
+            if (requiredVolumes === undefined) {
+                /*
+                 * Not need to continue, as this VM does not depend on any
```
Comment (misterdjules) [inline lib/workflows/nfs-volumes/index.js]: Good catch, will fix!

File: lib/common/validation.js [hunk 8]
```diff
@@ -908,6 +907,105 @@ function createValidateMarkerFn(field, options) {
...
+        var volume;
+        var volumes = params[fieldName];
+
+        if (required === false && volumes === undefined) {
+            return errs;
+        }
+
+        if (!Array.isArray(volumes)) {
+            errs.push(errors.invalidParamErr(fieldName,
+                    fieldName + ' must be an array of objects'));
```
Comment (misterdjules) [inline lib/common/validation.js]: No, you're right, I don't know how that and other lines of code ended up being over-indented. Will fix, thanks!

File: test/vms.volumes.test.js (hunk 1 above)
Comment (misterdjules) [inline test/vms.volumes.test.js]: Right, but two things:

1. Not setting a cap is _guaranteed_ to have issues, setting a cap has issues, as you pointed out, only in _some_ cases that we don't support when running Triton tests suites.

2. I assume we don't support a setup with both Triton and manta sharing provisionable CNs.

File: test/vms.volumes.test.js (hunk 2 above)
Comment (misterdjules) [inline test/vms.volumes.test.js]: Will fix.

File: test/vms.volumes.test.js (hunk 3 above)
Comment (misterdjules) [inline test/vms.volumes.test.js]: Will fix.

File: test/vms.volumes.test.js (hunk 4 above)
Comment (misterdjules) [inline test/vms.volumes.test.js]: Will fix.

File: test/vms.volumes.test.js (hunk 5 above)
Comment (misterdjules) [inline test/vms.volumes.test.js]: Will fix.

File: test/vms.volumes.test.js (hunk 6 above)
Comment (misterdjules) [inline test/vms.volumes.test.js]: Will fix.

**Comment by misterdjules** [comment]
Merged with 836aaa561b337441abc3b735f05c5693cb7f49c8.

**Pull request closed without merging**